### PR TITLE
[embedded-elt] Decoupling connections from resource allocation

### PR DESCRIPTION
**Pull request by tacastillo**
## Summary & Motivation

This PR proposes an alternative pattern to using Sling that uses a new Resource called `SlingConnectionResource` and builder method called `build_assets_from_sling_stream`. Names tentative.

Developers that use these objects can compose and mix-and-match the sources and targets interoperably by pushing the coupling down to the builder method. ie. in this snippet from a unit test, two different sources are synced to the same target database.

```python
sling_file_connection=SlingConnectionResource(type="file")
sling_staging_file_connection=SlingConnectionResource(type="file")
sling_sqlite_connection=SlingConnectionResource(type="sqlite", connection_string=f"sqlite://{temp_db}")
asset_def = build_assets_from_sling_stream(
    sling_file_connection,
    sling_sqlite_connection,
    stream=f"file://{test_csv}",
    target_object="main.tbl",
    mode=SlingMode.FULL_REFRESH,
    primary_key="SPECIES_CODE",
)

asset_def_two = build_assets_from_sling_stream(
    sling_staging_file_connection,
    sling_sqlite_connection,
    stream=f"file://{test_staging_csv}",
    target_object="main.staging_tbl",
    mode=SlingMode.FULL_REFRESH,
    primary_key="SPECIES_CODE",
)
```

## How I Tested These Changes
Unit tests

### CODE REVIEW

**Comment by tacastillo** [comment]
Current dependencies on/for this PR:
* master
  * **PR #18270** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/18270?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/18270?utm_source=stack-comment).

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/resources.py [hunk 1]
```diff
@@ -95,8 +96,127 @@ class SlingTargetConnection(PermissiveConfig):
     )
 
 
+class SlingSyncBase:
```
Comment (PedramNavid) [inline python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/resources.py:99]: Would add a docstring clarifying this class as well. probably worth making it @experimental

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/resources.py [hunk 2]
```diff
@@ -206,85 +305,116 @@ def _sync(
...
-                            source_connection=source,
-                            target_connection=target,
-                            mode=SlingMode.TRUNCATE,
-                        )
-                    },
-                )
+            # `_prep_cmd` only works with the Single Task command, so we need to replace it with the Replication command
+            # TODO: Unfortunately, _prep_cmd doesn't expose a way to add streams, so we'll need to override it ourselves
+            # We'll also need to parse the logs differently, since the output is different
+            # cmd = cmd.replace(" -c ", " -r ")
```
Comment (PedramNavid) [inline python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/resources.py:418]: Will a single task will still work with the new asset builder?

**Review by PedramNavid — APPROVED** [review]
Really nice work, just a few comments, no blockers.

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/resources.py (hunk 2 above)
Comment (tacastillo) [reply python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/resources.py:418]: Hmm...
In this PR, single task is what's used.
In the PR up the stack, we move the builder to only use replication.
We _could_ add a flag in the future so users can swap between them, with ack that the logic would be significantly different between the two.

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/resources.py (hunk 1 above)
Comment (tacastillo) [reply python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/resources.py:99]: `SlingSyncBase` isn't publicly exposed, should we still mark it experimental and document it?

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/asset_defs.py [hunk 3]
```diff
@@ -99,3 +104,61 @@ def sync(context: AssetExecutionContext) -> MaterializeResult:
...
+) -> AssetsDefinition:
+    if primary_key is not None and not isinstance(primary_key, list):
+        primary_key = [primary_key]
+
+    sling_replicator = SlingStreamReplicator(
+        source_connection=source,
+        target_connection=target,
+    )
+
+    asset_name = [stream]
```
[the remark this answers is not in the document]
Comment (tacastillo) [reply python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/asset_defs.py:129]: Yeah, I think I'll go back to what you did.

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/resources.py (hunk 1 above)
Comment (PedramNavid) [reply python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/resources.py:99]: If it's not a public class it should be marked `_SlingSyncBase`, but if you expect people might want to implement on top of this class then I would call add @experimental. 

In both cases, the docstring is helpful to explain the class's purpose to developers. I don't think you need the full API documentation with types/parms if its not public but 1-2 lines would be enough.

File: python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/resources.py (hunk 2 above)
Comment (tacastillo) [reply python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/sling/resources.py:418]: The comment was a placeholder (tbh, artifact) for myself for the PR that's upstack. Was trying to keep track of where the right place to swap out the single task for the replication.

'pologies for the confusion with the comment! For all intents and purposes, this PR uses single tasks.

**Pull request merged**